### PR TITLE
Implement reverseTransitionDuration for TransitionRoute

### PR DESCRIPTION
**Pull request by shihaohong**
## Description

This change allows developers to implement a reverse duration for a page transition. It defaults to the forward transitionDuration if it isn't specified.

```dart
// SAMPLE CLASS: 
class ModifiedReverseTransitionDurationRoute<T> extends MaterialPageRoute<T> {
  ModifiedReverseTransitionDurationRoute({
    @required WidgetBuilder builder,
    RouteSettings settings,
    Duration reverseTransitionDuration,
    bool fullscreenDialog = false,
  }) : _reverseTransitionDuration = reverseTransitionDuration,
       super(
         builder: builder,
         settings: settings,
         fullscreenDialog: fullscreenDialog,
       );

  @override
  Duration get reverseTransitionDuration => _reverseTransitionDuration;
  final Duration _reverseTransitionDuration;
}

// USAGE:
Navigator.of(context).push(
  ModifiedReverseTransitionDurationRoute<dynamic>(
    builder: (BuildContext innerContext) {
      return Container(color: Colors.green);
    },
    reverseTransitionDuration: const Duration(milliseconds: 150), // modified value
  ),
);
```

## Related Issues

n/a

## Tests

I added the following tests:

Added a test for the default reverse transition duration
Added a test for a custom reverse transition duration

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/routes.dart [hunk 1]
```diff
@@ -91,9 +91,12 @@ abstract class TransitionRoute<T> extends OverlayRoute<T> {
   Future<T> get completed => _transitionCompleter.future;
   final Completer<T> _transitionCompleter = Completer<T>();
 
-  /// The duration the transition lasts.
+  /// The duration the transition going forwards.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/routes.dart]: Maybe add a "see also" section to link to reverseTransitionDuration?

File: packages/flutter/lib/src/widgets/routes.dart [hunk 2]
```diff
@@ -91,9 +91,12 @@ abstract class TransitionRoute<T> extends OverlayRoute<T> {
...
   Duration get transitionDuration;
 
+  /// The duration the transition going in reverse.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/routes.dart]: Mention the default and link it to transitionDuraiton?

File: packages/flutter/test/widgets/routes_test.dart [hunk 3]
```diff
@@ -863,9 +863,131 @@ void main() {
...
+    bool fullscreenDialog = false,
+  }) : _reverseTransitionDuration = reverseTransitionDuration,
+       super(
+         builder: builder,
+         settings: settings,
+         fullscreenDialog: fullscreenDialog,
+       );
+
+  @override
+  Duration get reverseTransitionDuration => _reverseTransitionDuration;
```
Comment (goderbauer) [inline packages/flutter/test/widgets/routes_test.dart]: This and the next line could be simplified to:

```dart
@override
final Duration reverseTransitionDuration;
```

and then in the constructor you could just have `this.reverseTransitionDuration`.

**Comment by goderbauer** [comment]
What does it look like when you have a transitionDuration of x and a reverseTransitionDuration of y and while the push of a route is ongoing you pop it again? Is there a jump in the animation? Or does it smoothly switch direction?

**Comment by shihaohong** [comment]
When you interrupt the push of a route that is ongoing with a pop, it does not cause a jump in animation if the animations are properly defined. For example, with `ZoomPageTransitionsBuilder` and `FadeUpwardsTransitionsBuilder` smoothly switch directions. The only difference was that the reverse transition duration was honored.

**Review by goderbauer — APPROVED** [review]
LGTM

Maybe add a simple test to ensure that we never break the smooth animation when you pop while a push is ongoing with different durations?

**Comment by shihaohong** [comment]
@goderbauer I added the forward smooth transition test. PTAL

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**